drivers/hmis/app/models/hmis/project_group_criteria.rb

```diff
@@ -0,0 +1,168 @@
+    include ApplicationHelper
+    include ActionView::Helpers::TagHelper
+    include ActionView::Context
+    include ::Hmis::Concerns::HmisArelHelper
+
+    # Source data source for this project filter. ProjectGroupCriteria filter cannot be used across multiple data sources.
+    attribute :data_source_id, Integer, default: nil
+
+    # Attributes for filtering
+    attribute :project_ids, Array, default: []
+    attribute :organization_ids, Array, default: []
+    attribute :all_projects_in_data_source, Boolean, default: false
+    attribute :project_type_numbers, Array, default: []
+    # TODO: add more filtering capabilities:
+    # attribute :hmis_participation_status, Integer, default: nil
+    # attribute :ce_participation_access_point, Boolean, default: nil
+    # attribute :funder_ids, Array, default: []
+    # attribute :coc_codes, Array, default: []
+    # attribute :project_group_ids, Array, default: []
+    # attribute :coc_codes, Array, default: []
+    # attribute :project_status, String, default: 'all'
+
+    # Allowed attributes for validation and updates
+    ALLOWED_ATTRIBUTES = [
+      :project_ids,
+      :organization_ids,
+      :all_projects_in_data_source,
+      :project_type_numbers,
+      # :hmis_participation_status,
+      # :project_status,
+    ].freeze
+
+    # Validations
+    # validates :hmis_participation_status, inclusion: { in: [0, 1, 2], allow_nil: true, message: 'must be 0, 1, or 2' }
+    # validates :project_status, inclusion: { in: ['open', 'closed', 'all'], allow_nil: true, message: "must be 'open', 'closed', or 'all'" }
+
+    # Initialize with a hash or JSON blob
+    def initialize(input = {}, data_source_id:)
+      super()
+      update(input)
+      self.data_source_id = data_source_id
+    end
+
+    def update(attributes)
+      attributes = parse_input(attributes)
+      attributes.each do |key, value|
+        cleaned_value = value.is_a?(Array) ? value.reject(&:blank?) : value
+
+        raise ArgumentError, "Invalid attribute for ProjectGroupCriteria: #{key}. Allowed attributes are: #{ALLOWED_ATTRIBUTES.join(', ')}" unless ALLOWED_ATTRIBUTES.include?(key)
+
+        send("#{key}=", cleaned_value)
+      end
+      self
+    end
+
+    def to_h
+      ALLOWED_ATTRIBUTES.index_with { |attr| send(attr) }
+    end
+
+    def to_json(*_args)
+      to_h.to_json
+    end
+
+    def persisted?
+      false
+    end
+
+    # Returns a unique list of project IDs that match the criteria defined by this class.
+    def effective_project_ids
+      ids = []
+      ids << project_scope.where(id: project_ids).pluck(:id) if project_ids.any?
+      ids << organization_scope.joins(:projects).where(id: organization_ids).pluck(p_t[:id]) if organization_ids.any?
+      ids << project_scope.where(project_type: project_type_numbers).pluck(:id) if project_type_numbers.any?
+      ids << data_source.projects.pluck(p_t[:id]) if all_projects_in_data_source
+      ids.flatten.uniq
```

Comment:
I think we could just:
```suggestion
      return data_source.projects.pluck(p_t[:id]) if all_projects_in_data_source
      
      ids = []
      ids << project_scope.where(id: project_ids).pluck(:id) if project_ids.any?
      ids << organization_scope.joins(:projects).where(id: organization_ids).pluck(p_t[:id]) if organization_ids.any?
      ids << project_scope.where(project_type: project_type_numbers).pluck(:id) if project_type_numbers.any?
      ids.flatten.uniq
```
That's a micro performance optimization since you'll just have duplicates if they are all included.